[wasm] Webcil-in-WebAssembly

lambdageek opened the pull request:
Define a WebAssembly module wrapper for Webcil assemblies.
Contributes to #80807 

### Why

In some settings serving `application/octet-stream` data, or files with weird extensions will trigger firewalls or AV tools.  But let's assume that if you're interested in deploying a .NET WebAssembly app, you're in an environment that can at least serve WebAssembly modules.

### How

Essentially we serve this WebAssembly module:

```wat
(module
  (data "\0f\00\00\00") ;; data segment 0: payload size
  (data "webcil Payload\cc")  ;; data segment 1: webcil payload
  (memory (import "webcil" "memory") 1)
  (global (export "webcilVersion") i32 (i32.const 0))
  (func (export "getWebcilSize") (param $destPtr i32) (result)
    local.get $destPtr
    i32.const 0
    i32.const 4
    memory.init 0)
  (func (export "getWebcilPayload") (param $d i32) (param $n i32) (result)
    local.get $d
    i32.const 0
    local.get $n
    memory.init 1))
```

The module exports two WebAssembly functions `getWebcilSize` and `getWebcilPayload` that write some bytes (being the size or payload of the webcil assembly) to the linear memory at a given offset.  The module also exports the constant `webcilVersion` to version the wrapper format.

So a runtime or tool that wants to consume the webcil module can do something like:

```js
const wasmModule = new WebAssembly.Module (...);
const wasmMemory = new WebAssembly.Memory ({initial: 1});
const wasmInstance =
      new WebAssembly.Instance(wasmModule, {webcil: {memory: wasmMemory}});
const { getWebcilPayload, webcilVersion, getWebcilSize } = wasmInstance.exports;
console.log (`Version ${webcilVersion.value}`);
getWebcilSize(0);
const size = new Int32Array (wasmMemory.buffer)[0]
console.log (`Size ${size}`);
console.log (new Uint8Array(wasmMemory.buffer).subarray(0, 20));
getWebcilPayload(4, size);
console.log (new Uint8Array(wasmMemory.buffer).subarray(0, 20));
```

### How (Part 2)

But actually, we will define the wrapper to consist of exactly 2 data segments in the WebAssembly data section: segment 0 is 4 bytes and encodes the webcil payload size; and segment 1 is of variable size and contains the webcil payload.

So to load a webcil-in-wasm module, the runtime gets the _raw bytes_ of the WebAssembly module (ie: without instantiating it), and parses it to find the data section, assert that there are 2 segments, ensure they're both passive, and get the data directly from segment 1.

### Remaining work

- [x] Add an option to the Webcil converter to emit a wasm module containing the webcil assembly as a passive data segment
- [x] Add code to Mono's webcil loader to parse WASM modules and find the data segment
- [x] Update the webcil spec
- [x] Change the extension from ".webcil" to ".wasm" and update the wasm SDK assets logic and the runtime's extension probing logic
- [x] WebcilReader (used by DebugStore) needs ot understand webcil-in-wasm

lambdageek committed:
Add option to emit webcil inside a wasm module wrapper

lambdageek committed:
[mono][loader] implement a webcil-in-wasm reader

kg commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilConverter.cs:
```diff
@@ -62,6 +64,25 @@ public void ConvertToWebcil()
+
+            // webcil is about the same size as the PE file
+            using var memoryStream = new MemoryStream (checked((int)inputStream.Length));
+            WriteConversionTo(memoryStream, inputStream, peInfo, wcInfo);
```
Nit: inconsistent whitespace between method names and ( in here

kg commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilWasmWrapper.cs:
```diff
@@ -0,0 +1,166 @@
+// length of the webcil payload.  segment 1 is of a variable size and contains the webcil payload.
+//
+// the unchanging parts are stored as a "prefix" and "suffix" which contain the bytes for the following
+// WAT program, split into the parts that come before the data section, and the bytes that come after:
```
nit: i think in this case it's a WAT module; it's not a program since it isn't actually executable in a meaningful sense

lambdageek committed:
reword WebcilWasmWrapper summary comment

kg commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilWasmWrapper.cs:
```diff
@@ -0,0 +1,166 @@
+// WAT program, split into the parts that come before the data section, and the bytes that come after:
+//
+// (module
+//  (data "\0f\00\00\00") ;; data segment 0: payload size as a 4 byte LE uint32
```
I still don't understand why the payload size is in a data segment instead of in a global or the return value of a callable export, honestly

lambdageek replied:
because I don't want to edit a global section or a func section of a module

kg replied:
OK, so it's explicitly a zone that is safe to patch and is of a known size

lambdageek replied:
The data section is the easiest one to parse, so if we're going to do this "let the runtime read the bytes and pull out what it needs" thing, then it is ideal to keep everything variable in the data section.

To be honest the runtime doesn't really need the webcil payload size, so `webcilGetSize` could use a global.  But then I'd probably need to pull in a real WebAssembly parser/writer like wa-info, and since it hasn't yet landed in the runtime repo, I didn't want to be the one to do it 😁

lambdageek committed:
fix whitespace

lambdageek committed:
fix typos

kg commented on src/mono/mono/utils/wasm-module-reader.c:
```diff
@@ -0,0 +1,140 @@
+
+	gboolean stop = FALSE;
+	while (success && !stop && ptr < boundp) {
+		success = visit_section (ptr, boundp, &ptr, visitor, user_data, &stop);
```
this seems to implicitly rely on visit_section to advance to the end of the section, and visit_section seems to implicitly rely on section_visitor to do that. am I misunderstanding it? I think it would be ideal if visit_section always ensured that endp ended up in the right place

lambdageek replied:
oooh. good catch. actually I explicitly wanted `visit_section` to advance `ptr` and not to rely on the visitor (which is "user code" - and I don't actually want ot trust it to do the traversal).  There's a missing line in `visit_section` to bump `ptr` before returning

lambdageek committed:
visit_section should bump the ptr after traversal

pavelsavara commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilWasmWrapper.cs:
```diff
@@ -0,0 +1,168 @@
+        byte[]
+#endif
+        s_wasmWrapperPrefix = new byte[] {
+        0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, 0x01, 0x0a, 0x02, 0x60, 0x01, 0x7f, 0x00, 0x60, 0x02, 0x7f, 0x7f, 0x00, 0x02, 0x12, 0x01, 0x06, 0x77, 0x65, 0x62, 0x63, 0x69, 0x6c, 0x06, 0x6d,
```
This is problematic for source build. Perhaps there could at least be comment with link to source ?

lambdageek replied:
The source is in the comment above the class definition

lambdageek committed:
remove extra bytes from wasm webcil prefix

the split utility was including the data header prefix in the output

lambdageek committed:
don't forget to include number of segments in the data section

lambdageek committed:
update the Webcil spec to include the WebAssembly wrapper module

lambdageek committed:
fix typos and whitespace

lambdageek committed:
advance endp past the data segment payload

lambdageek committed:
Adjust RVA map offsets to account for wasm prefix

MonoImage:raw_data is used as a base when applying the RVA map to map
virtual addresses to physical offsets in the assembly.  With
webcil-in-wasm there's an extra wasm prefix before the webcil payload
starts, so we need to account for this extra data when creating the
mapping.

An alternative is to compute the correct offsets as part of generating
the webcil, but that would entangle the wasm module and the webcil
payload.  The current (somewhat hacky approach) keeps them
logically separate.

lambdageek committed:
Add a note about the rva mapping to the spec

lambdageek commented:
WBT tests should work now...

lambdageek committed:
Serve webcil-in-wasm as .wasm

lambdageek commented:
<img width="574" alt="Screenshot 2023-05-09 at 14 49 00" src="https://github.com/dotnet/runtime/assets/480437/5373b9fa-0a5f-4901-98ce-4e168971379d">

(The .webcil->.wasm commits are not... super-duper clean yet, but it seems like at least for simple cases everything is working...)

lambdageek commented:
Ok this is ready for review

lambdageek committed:
fix wbt

maraf commented on src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/ComputeWasmPublishAssets.cs:
```diff
@@ -576,7 +576,7 @@ private void ApplyPublishProperties(ITaskItem newAsemblyAsset)
 
             var extension = candidate.GetMetadata("Extension");
-            if (string.Equals(extension, ".dll", StringComparison.Ordinal) || string.Equals(extension, ".webcil", StringComparison.Ordinal))
+            if (string.Equals(extension, ".dll", StringComparison.Ordinal) || string.Equals(extension, ".webcil", StringComparison.Ordinal) || string.Equals (extension, Utils.WebcilInWasmExtension, StringComparison.Ordinal))
```
For blazor, we plan to support both `.wasm` and .`webcil` extensions, but in `WasmAppBuilder` we always use `.wasm`. Is that intended?

lambdageek committed:
remove old .webcil support from Sdk Pack Tasks

lewing committed:
Set SelfContained=true for browser-wasm runtimes (#86102)

lambdageek committed:
Implement support for webcil in wasm in the managed WebcilReader

lambdageek commented:
debugger tests are working in manual testing.

next steps:

- [ ] figure out why the icu sharding WBT tests aren't working on CI
- [ ] add a CI leg that runs debugger tests with webcil

lambdageek committed:
why fail?

lambdageek committed:
did we load the same asm twice??

lambdageek commented:
I think I somehow managed to make the runtime load both "System.Private.CoreLib.dll" and "System.Private.CoreLib.wasm"... otherwise I don't understand how despite a `catch (CultureNotFoundException)`, I'm getting an uncaught `CultureNotFoundException`...

pavelsavara commented:
```Error: Culture is not supported. 
fr-FR is an invalid culture identifier.
```

I wonder is this is related to ICU data files or to some satellite assemblies ?

ilonatommy commented:
> ```
> fr-FR is an invalid culture identifier.
> ```
> 
> I wonder is this is related to ICU data files or to some satellite assemblies ?

ICU data files are loaded fine in the test. We are loading custom ICU file that does not contain fr-FR. We are also setting `PredefinedCulturesOnly` to true - these two things should result in `CultureNotFoundException` when we try to create fr-FR cultureand they do, we get `fr-FR is an invalid culture identifier.`, but it seems we are not getting into the catch block, so exception type must have changed. I'll edit the tests to see where it comes from

lambdageek commented:
> but it seems we are not getting into the catch block, so exception type must have changed

As far as I can tell it's the same exception as before.  I think what is happening is that the webcil packaging is somehow allowing the runtime to load the same assembly (CoreLib) twice.  So somehow the `CultureNotFoundException` from the `throw`, and the `CultureNotFoundException` from the `catch` are from two different copies of CoreLib.  And so the runtime can't match them up.

I think I see how it is happening. I'm going to try to repro locally

lambdageek commented:
> I think I see how it is happening. I'm going to try to repro locally

I can repro locally, but I think I was wrong about what is going wrong - or at least i'm getting a different-seeming failure locally.

I think getting resource streams from assemblies isn't working right when webcil-in-wasm (it does appear to work right with standalone .webcil.  So it's something about offsets into a `MonoImage:raw_data` - or else wrapping the image in .wasm is being done incorrectly and i'm truncating the image somehow...

lambdageek committed:
checkpoint. things are broken. but I adjusted MonoImage:raw_data

lambdageek committed:
Merge remote-tracking branch 'origin/main' into webcil-wasm-wrapper

lambdageek commented:
I think I understand what the issue was.  It wasn't resources (although there was an issue with resources) and it wasn't somehow loading the same assembly multiple times.

The issue is alignment.

When we load webcil-in-wasm by reading the wasm module directly, the webcil payload wasn't aligning on a good bounadry.  The binary WASM format doesn't really take care to align data segments on multi-byte boundaries - and the uleb128 encoding of all the internal data sizes would make that hard to achieve, anyway.  On the other hand, webcil and PE do end up with various internals on 4-byte boundaries.

As a result, the code for parsing a the exception clause information for an ECMA-335 method header was doing a call to align the input pointer to a 4-byte boundary and ending up at an unexpected offset.  Which resulted in mis-parsing the header and not setting up a correct catch clause for the `CultureNotFoundException` when webcil-in-wasm was turned on.

https://github.com/dotnet/runtime/blob/63e677701d35c7e2dad4fe74a84eb935cd396155/src/mono/mono/metadata/metadata.c#L4157-L4158


The solution is to change the spec slightly and use data segment 0 not only to carry the size of the webcil payload, but also to add extra padding bits so that the content of data segment 1 (not segment 1's header!) ends up on a 16-byte boundary (smaller is probably ok, but 16 is probably a safer choice).

(We don't want to somehow pad out segment 1's data because in the case where we _do_ use module instantiation then we wouldn't want the segment 1 data to be extracted by `getWebcilPayload` to an address with a proper alignment.)

Adding padding to segment 0 is ok because we only want the first 4 bytes of it anyway.

lambdageek committed:
align webcil payload to a 4-byte boundary within the wasm module

Add padding to data segment 0 to ensure that data segment 1's
payload (ie the webcil content itself) is 4-byte aligned

lambdageek commented:
pushed a commit to fix the alignment problem.  also pushed some debugging junk, so that will need to be backed out before we merge.  but I want to see how CI does.

Also there's a fix here to ensure the debugger sees the debugging sections of the webcil image.

The upshot is that for in-memory MonoImageStorage, the webcil loader will bump `MonoImage:raw_data` to the start of the payload - otherwise the PDB data's embedded offset pointers (which count from the beginning of the "PE" file are wrong).  To de-allocate the image when we close, we use the existing `MonoImage:raw_data_handle` field (which was previously only used for mmap handles and such) to point to the beginning of the `malloc`'d memory and pass it to `free`.

lambdageek committed:
remove WIP tracing

lambdageek committed:
assert that webcil raw data is 4-byte aligned

lambdageek committed:
revert unrelated build change

lambdageek committed:
revert unrelated change

lambdageek committed:
revert whitespace

lambdageek committed:
revert WBT debugging output changes

lambdageek committed:
add 4-byte alignment requirement to the webcil spec

lambdageek committed:
Don't modify MonoImageStorage:raw_data

instead just keep track of the webcil offset in the MonoImageStorage.

This introduces a situation where MonoImage:raw_data is different from
MonoImageStorage:raw_data.  The one to use for accessing IL and
metadata is MonoImage:raw_data.

The storage pointer is just used by the image loading machinery

lambdageek commented:
WBT is passing 🎉 

but over in the other PR #86330 some debugger tests are failing, I want to get those fixes in here before merging this PR .  (I couldn't figure out how to add a new variant of the debugger runs yet - so I'm testing webcil-in-wasm debugger support by changing the default packaging and running the "normal" debugger lane).

lambdageek commented:
Stepping tests were failing with plain .webcil also.  it's a problem in the debugger's webcil parser where it is calling `BlobReader.ReadUtf8NullTerminated` using reflection and failing to pick up the updated state (BlobReader is a struct type).  I'll fix it in the "on by default" PR.

This one is :shipit:
